java-frontend/src/main/java/org/sonar/java/se/constraint/ConstraintsByDomain.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.se.constraint;
+
+import org.sonar.java.collections.PCollections;
+import org.sonar.java.collections.PMap;
+
+import javax.annotation.Nullable;
+
+import java.util.function.BiConsumer;
+import java.util.stream.Stream;
+
+public class ConstraintsByDomain {
+
+  private final PMap<Class<? extends Constraint>, Constraint> constraintPMap;
+
+  private static final ConstraintsByDomain EMPTY = new ConstraintsByDomain(PCollections.emptyMap());
+
+  private ConstraintsByDomain(PMap<Class<? extends Constraint>, Constraint> constraintPMap) {
+    this.constraintPMap = constraintPMap;
+  }
+
+  public static ConstraintsByDomain empty() {
+    return EMPTY;
+  }
+
+  public ConstraintsByDomain remove(Class<? extends Constraint> domain) {
+    PMap<Class<? extends Constraint>, Constraint> remove = constraintPMap.remove(domain);
```

Comment:
Similarly to the `put` method, I think it would be more judicious to check if after the remove operation the new PMap is similar to the previous one and returning it directly if it's the case, rather than systematically create a new on. That would correspond to a case of removing a domain which is not part of the current domains.